shared/js/background/newtab-tracker-stats.js

```diff
@@ -0,0 +1,348 @@
+                urls: ['<all_urls>']
+            },
+            ['blocking']
+        )
+
+        /**
+         * Handle runtime messages sent from the new tab page
+         */
+        browser.runtime.onMessage.addListener((event, sender) => {
+            if (sender.id !== browserWrapper.getExtensionId()) return
+            this.handleIncomingEvent(event)
+        })
+
+        /**
+         * Register an alarm and handle when it fires.
+         * For now we're pruning data every 5 min
+         */
+        const pruneAlarmName = 'pruneNewTabData'
+        browserWrapper.createAlarm(pruneAlarmName, { periodInMinutes: 5 })
+        browser.alarms.onAlarm.addListener(async alarmEvent => {
+            if (alarmEvent.name === pruneAlarmName) {
+                this.handlePruneAlarm()
+            }
+        })
+
+        /**
+         * listen for the 'tracker-blocked' event that is fired from `before-request.es6.js`
+         * when a request is either blocked or a surrogate was used
+         */
+        emitter.on(TrackerBlockedEvent.eventName, (event) => {
+            if (!(event instanceof TrackerBlockedEvent)) return
+            this.record(event.companyDisplayName)
+        })
+
+        /**
+         * Assign the entities from the `tds` data
+         */
+        this.assignTopCompanies(tdsStorage.tds.entities)
+    }
+
+    /**
+     * @param {Record<string, { displayName: string, prevalence: number }>} entities
+     * @param {number} [maxCount] - how many to consider 'top companies'
+     */
+    assignTopCompanies (entities, maxCount = 100) {
+        const sorted = Object.keys(entities)
+            .map(
+                /** @returns {[string, number]} */
+                (key) => {
+                    const current = entities[key]
+                    return [current.displayName, current.prevalence]
+                })
+            .sort((a, b) => b[1] - a[1])
+            .slice(0, maxCount)
+
+        this.top100Companies = new Map(sorted)
+    }
+
+    /**
+     * This will be called for every 'blocked' event. It can be called many times
+     * and in quick succession, hence the debouncing
+     *
+     * @param {string} displayName
+     * @param {number} [timestamp] - optional timestamp
+     */
+    record (displayName, timestamp) {
+        /**
+         * Increment the count of this company if the following 2 predicates are satisfied
+         *
+         * 1) the `displayName` of the company is in our top100Companies list
+         * 2) the `displayName` of the company is NOT in our `excludedCompanies` list
+         */
+        if (this.top100Companies?.has(displayName) && !constants.trackerStats.excludedCompanies.includes(displayName)) {
+            this.stats.increment(displayName, timestamp)
+        } else {
+            /**
+             * Otherwise just increase the 'unknown' count
+             */
+            this.stats.increment(NewTabTrackerStats.unknownCompanyKey, timestamp)
```

Comment:
Shouldn't we call this 'others' instead of 'unknown'? Should we differentiate between actual 'unknown' trackers (where `findParentDisplayName` returns 'unknown'), and known by not top100 trackers?

Reply:
Yep, done :)